src/daemon/distribution.c

```diff
@@ -0,0 +1,262 @@
+  double new_max;
+
+  for (size_t i = 0; i < num_buckets; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] = initialize_bucket(0, base);
+    } else if (i < num_buckets - 1) {
+      /*previous_bucket_size = new_distribution->buckets[i - 1].max_boundary -
+                             new_distribution->buckets[i - 1].min_boundary + 1;
+      new_max = new_distribution->buckets[i - 1].max_boundary +
+                previous_bucket_size * factor;*/
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, base * pow(factor, i));
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets = num_buckets;
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+distribution_t *distribution_new_custom(size_t num_bounds,
+                                        double *custom_max_boundaries) {
+
+  if ((num_bounds == 0) || (custom_max_boundaries == NULL)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  for (size_t i = 1; i < num_bounds; i++) {
+    if (custom_max_boundaries[i] <= custom_max_boundaries[i - 1]) {
+      errno = EINVAL;
+      return NULL;
+    }
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets =
+      calloc(num_bounds + 1, sizeof(bucket_t)); //+1 for infinity bucket
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_bounds + 1; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] =
+          initialize_bucket(0, custom_max_boundaries[0]);
+    } else if (i < num_bounds) {
+      new_distribution->buckets[i] =
+          initialize_bucket(new_distribution->buckets[i - 1].max_boundary,
+                            custom_max_boundaries[i]);
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets =
+      num_bounds + 1; // plus one for infinity bucket
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+static size_t binary_search(distribution_t *dist, size_t left, size_t right,
+                            double gauge) {
+  if (left > right) {
+    return -1;
+  }
+
+  size_t mid = left + (right - left) / 2;
+  if (gauge >= dist->buckets[mid].min_boundary &&
+      gauge < dist->buckets[mid].max_boundary) {
+    return mid;
+  }
+
+  if (gauge < dist->buckets[mid].min_boundary) {
+    return binary_search(dist, left, mid, gauge);
+  }
+
+  return binary_search(dist, mid, right, gauge);
+}
+
+void distribution_update(distribution_t *dist, double gauge) {
+
```

Comment:
Please check that `dist != NULL` and `gauge > 0`.

Reply:
Okay.